File: src/main/kotlin/sunsetsatellite/vm/sunlite/SLArray.kt
package sunsetsatellite.vm.sunlite

import sunsetsatellite.lang.sunlite.Sunlite

class SLArray(var size: Int, val sunlite: Sunlite) {
	private var array: Array<AnySLValue> = VM.arrayOfNils(size)

	fun internal(): Array<AnySLValue> {
		return array
	}

	fun overwrite(arr: Array<AnySLValue>): SLArray {
		array = arr
		return this
	}

	fun set(index: Int, value: AnySLValue) {
		if(index >= size) {
			sunlite.vm.runtimeError("Array index $index is out of bounds for an array of size $size.")
		}
		array[index] = value
	}

	fun get(index: Int): AnySLValue {
		if(index >= size) {
			sunlite.vm.runtimeError("Array index $index is out of bounds for an array of size $size.")
		}
		return array[index]
	}

	fun resize(newSize: Int) {
		val newArray = VM.arrayOfNils(newSize)
		System.arraycopy(array, 0, newArray, 0, array.size.coerceAtMost(newSize))
		size = newSize
		array = newArray
	}

	fun copy(): SLArray {
		return SLArray(size, sunlite).overwrite(array.map { it.copy() }.toTypedArray())
	}

	override fun toString(): String {
		return "<array of size $size>"
	}
}

File: src/main/csharp/SunliteSharp/Runtime/SLArray.cs
using System.Drawing;

namespace SunliteSharp.Runtime;

public class SLArray(int size, Sunlite sl)
{
    private AnySLValue[] _values = VM.ArrayOfNils(size);

    public AnySLValue this[int index]
    {
        get
        {
            if (index < size) return _values[index];
            sl.Vm.RuntimeError($"Array index {index} is out of bounds for array of size {size}.");
            return SLNil.Nil;
        }
        set
        {
            if (index < size) _values[index] = value;
            sl.Vm.RuntimeError($"Array index {index} is out of bounds for array of size {size}.");
        }
    }
    
    public AnySLValue[] Internal => _values;

    public void Resize(int newSize)
    {
        Array.Resize(ref _values, newSize);
        _values = _values.Concat(VM.ArrayOfNils(newSize - size)).ToArray();
    }

    public SLArray Overwrite(AnySLValue[] arr)
    {
        _values = arr;
        return this;
    }

    public SLArray Copy()
    {
        return new SLArray(size, sl).Overwrite(_values.Select(v => v.Copy()).ToArray());
    }

    public override string ToString()
    {
        return $"<array of size {size}>";
    }
}